LogAnalysis.hs

```diff
@@ -0,0 +1,101 @@
+insert lmX (Node t lmY s)
+    | getInt lmX <= getInt lmY = Node (insert lmX t) lmY s  -- step case, less than or equal
+    | getInt lmX > getInt lmY  = Node t lmY (insert lmX s)  -- step case, opposite to above 
+insert (Unknown _) (Node t lmY s) = Node t lmY s  -- Unknowns should not affect a messageTree 
+
+---------------------------------------------------------------------------------
+-- Exercise 3: build
+---------------------------------------------------------------------------------
+-- Makes a messageTree from a list of messages 
+
+-- Getting timestamps from a list of log messages 
+getStamps :: [LogMessage] -> [TimeStamp]
+getStamps [] = []
+getStamps (lmX:lmList) = [getInt lmX] ++ getStamps lmList
+
+build :: [LogMessage] -> MessageTree
+build [] = Leaf  -- Takes empty list, produces empty node 
+-- left subtree has half the list, right subtree has the other half 
+build (lmX:lmList) = Node 
+                     (build (filter (\msg -> getInt msg < getInt lmX) lmList)) lmX
+                     (build (filter (\msg -> getInt msg > getInt lmX) lmList)) 
+
+---------------------------------------------------------------------------------
+-- Exercise 4: inOrder
+---------------------------------------------------------------------------------
+inOrder :: MessageTree -> [LogMessage]
+inOrder Leaf = []  -- No log messages in an empty node, so empty list generated 
+inOrder (Node t lmY s) = inOrder t ++ [lmY] ++ inOrder s  -- Recursively makes list 
+
+---------------------------------------------------------------------------------
+-- Exercise 5: whatWentWrong
+---------------------------------------------------------------------------------
+-- Takes unsorted list of LogMessages, returns list of messages of any errors with 
+-- severity of 50 or more 
+
+-- Function to get severity of all kinds of logMessages, default for non-errors is 0
+getSevere :: LogMessage -> Int 
+getSevere (LogMessage (Error b) _ _) = b
+getSevere (LogMessage _ a _) = 0
+getSevere (Unknown _) = 0
+
+-- building a sorted list of logMessages, using build and then inOrder 
+sortList :: [LogMessage] -> [LogMessage]
+sortList lmList = inOrder $ build lmList
+
+-- Getting the actual description of sever error LogMessages 
+getString :: [LogMessage] -> String
+getString [(LogMessage (Error _) _ s)] = s  -- Order matters in specifying how to get strings
+getString [(Unknown _)] = ""  -- Say empty here because we are not interested in non-errors
+getString [(LogMessage _ _ _)] = "" 
+getString [] = ""
+
+-- Extract messages of type error, of whose severity >= 50
+bad :: [LogMessage] -> [LogMessage]
+bad lmList = filter (\msg -> getSevere msg >= 50) lmList
+
+-- Putting the descritptions in a new list 
+whatWentWrong :: [LogMessage] -> [String]
+whatWentWrong [] = []
+-- Filter out all non-error messages based on their appropriate strings defined above 
+whatWentWrong (lmX:lmList) = filter (/="") (getString (bad [lmX]) : whatWentWrong (bad lmList))
```

Comment:
I would recommend to implement this function using standard `map` and `filter` function. So you can write just:

```haskell
whatWentWrong msgs = map extractMessage $ filter isHighSevery msgs
```

You only need to implement `extractText` and `isHighSevery` functions. Also, as I can see from task description, you need to use sorting according to timestamps here.